internal/linker/link_node_modules.ts

```diff
@@ -90,49 +90,71 @@ async function symlink(target: string, p: string): Promise<boolean> {
 }
 
 /**
- * Resolve a root directory string to the actual location on disk
- * where node_modules was installed
- * @param root a string like 'npm/node_modules'
+ * Resolve to an absolute root node_modules directory.
+ * @param root The bazel managed node_modules root such as 'npm/node_modules',  which includes the
+ * workspace name as the first segment. May be undefined if there are no third_party node_modules
+ * deps.
+ * @param startCwd The absolute path that bazel started the action at.
+ * @param isExecroot True if the action is run in the execroot, false if the action is run in
+ * runfiles root.
+ * @param runfiles The runfiles helper object.
+ * @return The absolute path on disk where node_modules was installed or if no third party
+ * node_modules are deps of the current target the returns the absolute path to
+ * `execroot/my_wksp/node_modules`.
  */
-async function resolveRoot(root: string|undefined, runfiles: Runfiles) {
-  if (!runfiles.execroot) {
-    // Under runfiles, the repository should be layed out in the parent directory
-    // since bazel sets our working directory to the repository where the build is happening
-    process.chdir('..');
+async function resolveRoot(
+    root: string|undefined, startCwd: string, isExecroot: boolean, runfiles: Runfiles) {
+  if (isExecroot) {
+    // Under execroot, the root will be under an external folder from the startCwd
+    // `execroot/my_wksp`. For example, `execroot/my_wksp/external/npm/node_modules`. If there is no
+    // root, which will be the case if there are no third-party modules dependencies for this
+    // target, set the root to `execroot/my_wksp/node_modules`.
+    return root ? `${startCwd}/external/${root}` : `${startCwd}/node_modules`;
   }
-  // create a node_modules directory if no root
-  // this will be the case if only first-party modules are installed
+
+  // Under runfiles, the linker should symlink node_modules at `execroot/my_wksp`
+  // so that when there are no runfiles (default on Windows) and scripts run out of
+  // `execroot/my_wksp` they can resolve node_modules with standard node_module resolution
+
+  // Look for bazel-out which is used to determine the the path to `execroot/my_wksp`. This works in
+  // all cases including on rbe where the execroot is a path such as `/b/f/w`. For example, when in
+  // runfiles on rbe, bazel runs the process in a directory such as
+  // `/b/f/w/bazel-out/k8-fastbuild/bin/path/to/pkg/some_test.sh.runfiles/my_wksp`. From here we can
+  // determine the execroot `b/f/w` by finding the first instance of bazel-out.
+  const match = startCwd.match(/\/bazel-out\//);
+  if (!match) {
+    panic('No bazel-out folder found in path!')
```

Comment:
nit: include more info about what path we looked in?

Reply:
👍